src/EventStore.Core/TransactionLog/Scavenging/Stages/StreamCalculator.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using EventStore.Core.Data;
+
+namespace EventStore.Core.TransactionLog.Scavenging {
+	// the idea of this is we load some context into this class for the given stream
+	// and then it helps us calculate and make decisions about what to keep.
+	// and the calculations can be done on demand.
+	// and can be unit tested separately if necessary
+	// reused between streams to avoid allocations.
+	public class StreamCalculator<TStreamId> {
+		public StreamCalculator(
+			IIndexReaderForCalculator<TStreamId> index,
+			ScavengePoint scavengePoint) {
+
+			Index = index;
+			ScavengePoint = scavengePoint;
+		}
+
+		public void SetStream(
+			StreamHandle<TStreamId> originalStreamHandle,
+			OriginalStreamData originalStreamData) {
+
+			_lastEventNumber = null;
+			_truncateBeforeOrMaxCountDiscardPoint = null;
+
+			OriginalStreamHandle = originalStreamHandle;
+			OriginalStreamData = originalStreamData;
+		}
+
+		// State that doesn't change. scoped to the scavenge.
+		public IIndexReaderForCalculator<TStreamId> Index { get; }
+		public ScavengePoint ScavengePoint { get; }
+
+		// State that is scoped to the stream
+		public StreamHandle<TStreamId> OriginalStreamHandle { get; private set; }
+		private OriginalStreamData OriginalStreamData { get; set; }
+
+		// Returns NoStream (-1) if there are no events before the scavenge point.
+		// Caller must handle that
+		private long? _lastEventNumber;
+		public long LastEventNumber {
+			get {
+				if (!_lastEventNumber.HasValue) {
+					_lastEventNumber = Index.GetLastEventNumber(OriginalStreamHandle, ScavengePoint);
+				}
+
+				return _lastEventNumber.Value;
+			}
+		}
+
+		public DiscardPoint TruncateBeforeDiscardPoint =>
+			OriginalStreamData.TruncateBefore.HasValue
+				? DiscardPoint.DiscardBefore(OriginalStreamData.TruncateBefore.Value)
+				: DiscardPoint.KeepAll;
+
+		public DiscardPoint MaxCountDiscardPoint =>
+			// if LastEventNumber is NoStream (-1) this will always KeepAll as intended
+			OriginalStreamData.MaxCount.HasValue
+				? DiscardPoint.DiscardIncluding(LastEventNumber - OriginalStreamData.MaxCount.Value)
+				: DiscardPoint.KeepAll;
+
+		// accounts for tombstone, truncateBefore, maxCount.
```

Comment:
```suggestion
		// accounts for truncateBefore, maxCount.
```

Reply:
typo?

Reply:
yep, ive removed the comment since it turns out to be self explanatory anyway

Reply:
ok :+1: